x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java

```diff
@@ -79,6 +86,23 @@ private String getEnrichIndexKey(SearchRequest searchRequest) {
         return ia.getIndices().get(0).getIndex().getName();
     }
 
+    CompletableFuture<SearchResponse> computeIfAbsent(
```

Comment:
What about adding this method instead of these two methods:

```
void computeIfAbsent(
        SearchRequest searchRequest,
        BiConsumer<SearchRequest, ActionListener<SearchResponse>> executeSearch,
        BiConsumer<SearchResponse, Exception> handler
    ) {
        String enrichIndex = getEnrichIndexKey(searchRequest);
        CacheKey cacheKey = new CacheKey(enrichIndex, searchRequest);

        try {
            CompletableFuture<SearchResponse> cacheEntry = cache.computeIfAbsent(cacheKey, request -> {
                CompletableFuture<SearchResponse> completableFuture = new CompletableFuture<>();
                ActionListener<SearchResponse> listener = ActionListener.wrap(
                    completableFuture::complete,
                    completableFuture::completeExceptionally
                );
                executeSearch.accept(searchRequest, listener);
                return completableFuture;
            });
            cacheEntry.whenComplete((response, throwable) -> {
                Exception exception = null;
                if (throwable != null) {
                    cache.invalidate(cacheKey, cacheEntry);
                    if (throwable instanceof Exception) {
                        exception = (Exception) throwable;
                    } else {
                        exception = new Exception("Unexpected error", throwable);
                    }
                }
                handler.accept(response, exception);
            });
        } catch (ExecutionException e) {
            handler.accept(null, e);
        }
    }
```

and then `EnrichProcessorFactory#createSearchRunner()` can look like this:

```
Client originClient = new OriginSettingClient(client, ENRICH_ORIGIN);
        return (req, handler) -> {
            enrichCache.computeIfAbsent(
                req,
                (searchRequest, listener) -> originClient.execute(EnrichCoordinatorProxyAction.INSTANCE, searchRequest, listener),
                handler
            );
        };
```

This way the logic of dealing with `Futures` is kept inside the cache and then there is also no need to expose the `invalidate(...)` method.

Reply:
(small nit: maybe move this method up to the other methods with package protected visibility)

Reply:
Pulling the logic up makes sense, will give it a go.